### ISSUE TITLE
2 hit troll not repairing at end or beginning of turn

### ISSUE THREAD

**Issue by justbleh** [triplea-game/triplea#2600]
### TripleA version:
1.9.0.0.7509
[dwarves vs the world hard.zip](https://github.com/triplea-game/triplea/files/1464420/dwarves.vs.the.world.hard.zip)

### Map:
Large Mordor map
### Can you describe how to trigger the error? (eg: what sequence of actions will recreate it?)
go into battle and 2 turns later none of my 2 hit characters have healed
### Do you have the exact error text? Please copy/paste if so
no
### Instead of this error, what should have happened?
they should have healed
### Any additional information that may help:
[Angmar vs the world hard.zip](https://github.com/triplea-game/triplea/files/1464421/Angmar.vs.the.world.hard.zip)

**Comment by justbleh** [comment]
this error message keeps popping up as well.
triplea.engine.version.bin:1.9
Loading map: large_middle_earth, from: C:\Users\johnw_000\triplea\downloadedMaps\large_middle_earth-master.zip
Loading resources from the following paths: [C:\Users\johnw_000\triplea\downloadedMaps\large_middle_earth-master.zip, C:\Program Files\TripleA\assets]
Loading map: large_middle_earth, from: C:\Users\johnw_000\triplea\downloadedMaps\large_middle_earth-master.zip
Loading resources from the following paths: [C:\Users\johnw_000\triplea\downloadedMaps\large_middle_earth-master.zip, C:\Program Files\TripleA\assets]
java.lang.reflect.UndeclaredThrowableException
	at com.sun.proxy.$Proxy13.selectCasualties(Unknown Source)
	at games.strategy.triplea.delegate.BattleCalculator.selectCasualties(BattleCalculator.java:514)
	at games.strategy.triplea.delegate.Fire.selectCasualties(Fire.java:161)
	at games.strategy.triplea.delegate.Fire.access$100(Fire.java:21)
	at games.strategy.triplea.delegate.Fire$2.execute(Fire.java:221)
	at games.strategy.triplea.delegate.ExecutionStack.execute(ExecutionStack.java:41)
	at games.strategy.triplea.delegate.MustFightBattle.fight(MustFightBattle.java:380)
	at games.strategy.triplea.delegate.BattleTracker.fightBattleIfOnlyOne(BattleTracker.java:1176)
	at games.strategy.triplea.delegate.BattleDelegate.start(BattleDelegate.java:97)
	at games.strategy.engine.framework.ServerGame.startStep(ServerGame.java:506)
	at games.strategy.engine.framework.ServerGame.runStep(ServerGame.java:401)
	at games.strategy.engine.framework.ServerGame.startGame(ServerGame.java:253)
	at games.strategy.engine.framework.startup.launcher.LocalLauncher.launchInNewThread(LocalLauncher.java:61)
	at games.strategy.engine.framework.startup.launcher.AbstractLauncher.lambda$launch$169(AbstractLauncher.java:51)
	at java.lang.Thread.run(Unknown Source)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at games.strategy.engine.delegate.DelegateExecutionManager.lambda$createOutboundImplementation$113(DelegateExecutionManager.java:116)
	... 15 more
Caused by: games.strategy.engine.message.RemoteNotFoundException: Not found:games.strategy.engine.framework.ServerGame.PLAYER_REMOTE.Neutral
	at games.strategy.engine.message.UnifiedMessengerHub.messageReceived(UnifiedMessengerHub.java:89)
	at games.strategy.engine.message.unifiedmessenger.UnifiedMessenger.send(UnifiedMessenger.java:224)
	at games.strategy.engine.message.unifiedmessenger.UnifiedMessenger.invokeAndWaitRemote(UnifiedMessenger.java:122)
	at games.strategy.engine.message.unifiedmessenger.UnifiedMessenger.invokeAndWait(UnifiedMessenger.java:99)
	at games.strategy.engine.message.UnifiedInvocationHandler.invoke(UnifiedInvocationHandler.java:51)
	at com.sun.proxy.$Proxy13.selectCasualties(Unknown Source)
	... 20 more
Caused by: java.lang.IllegalStateException: remote not found
	at java.lang.Thread.getStackTrace(Unknown Source)
	at games.strategy.engine.message.MessengerException.fillInInvokerStackTrace(MessengerException.java:19)
	at games.strategy.engine.message.UnifiedInvocationHandler.invoke(UnifiedInvocationHandler.java:55)
	... 21 more

**Comment by RoiEXLab** [comment]
According to your error message this seems related to https://github.com/triplea-game/triplea/issues/2594
Could you verify this issue is still present in the lates pre-release?

**Comment by ssoloff** [comment]
@justbleh 

> go into battle and 2 turns later none of my 2 hit characters have healed

I am unable to reproduce the repair issue in 1.9.0.0.7522 using the _dwarves.vs.the.world.hard.zip_ save game you provided.

Using edit mode, I gave the Orcs trolls in Celebdil and Limlight one hit each:

![trolls-with-one-hit](https://user-images.githubusercontent.com/4826349/32710357-87db488e-c805-11e7-8e16-eaa86e604830.png)

At the end of the Orcs turn, those trolls were repaired:

![trolls-after-repair](https://user-images.githubusercontent.com/4826349/32710364-93952456-c805-11e7-9734-4357846f07e7.png)

I'm probably missing something in the repro steps.  Can you please point out the specific units that are not being repaired?  It may be something that's only affecting one of the players (and the Orcs example I used may not be one of those players).

> this error message keeps popping up as well.

As @RoiEXLab pointed out, that's the same error you reported in #2594, which should be fixed in 1.9.0.0.7509 and later.  Using both of the save games you provided, I never saw this error occur.

However, using _Angmar.vs.the.world.hard.zip_, I did consistently see the following error during the battle at which the game is saved at:

```
Current battle exists but is not on pending list:  Battle in:West Rhudaur battle type:Battle defender:Freefolk attacked by:Angmar from:[Weathertop, South Angmar] attacking with: []
```

Is that what you're talking about?

**Comment by justbleh** [comment]
The error message keeps popping up in a new message, so I added that to this issue. I don't know what it means at all. The main unit is the dragon and troll in Angmar in the saved Angmar map. I didn't use them for two rounds and didn't get attacked and they never healed. In battle a few turns later one hit kills them, then I just enter edit mode and add another....

**Comment by ssoloff** [comment]
> The main unit is the dragon and troll in Angmar in the saved Angmar map.

At the point of the _Angmar.vs.the.world.hard.zip_ save game (combat phase of Angmar's turn in round 4), the state of Angmar's damaged multi-hit units is the following:

Territory | Unit | Hits
:-- | :-- | :--
West Angmar | Dragon | 2 of 3
West Rhudaur | Troll (a) | 1 of 2
West Rhudaur | Troll (b) x2 | 1 of 2

(Note that I don't understand what's different between the one Troll (a) unit and two Troll (b) units in West Rhudaur that causes the engine to stack them separately.  But all three appear to have 1 of 2 hits.)

At the end of Angmar's turn in round 4, the history states that 4 units were repaired, and the state of the units in question is the following:

Territory | Unit | Hits
:-- | :-- | :--
West Angmar | Dragon | 1 of 3
West Rhudaur | Troll (a) | 1 of 2
West Rhudaur | Troll (b) x2 | 0 of 2

:warning: I'm not sure why Troll (a) was reported as being repaired, but the hit count remained at 1.

At the end of Angmar's turn in round 5, the history states that 2 units were repaired, and the state of the units in question is the following:

Territory | Unit | Hits
:-- | :-- | :--
West Angmar | Dragon | 0 of 3
West Rhudaur | Troll x3 | 0 of 2

Note at this point, the engine merges all three trolls in West Rhudaur into a single stack.

So, it appears the Angmar units are ultimately fully repaired if left alone.  The only issue I observed is why the Troll (a) unit in West Rhudaur did not have its hit count decremented at the end of round 4 even though the engine reported that that unit was repaired.

@ron-murhammer @panther2 @FrostionAAA: If any of you have time, could you please take a quick look at the _Angmar.vs.the.world.hard.zip_ save to see if there's an obvious reason why the Troll (a) unit in West Rhudaur (that attacked from Weathertop) is not fully repaired at the end of round 4?

**Comment by justbleh** [comment]
Thanks for all of your help with this. I also have some thoughts on the AI for this map, who would I contact to discuss them with?


________________________________

**Comment by ron-murhammer** [comment]
@ssoloff @justbleh The issue seems to be before that battle in West Rhudaur even occurs. The Troll in Weathertop actually has 0 HP (you can check by opening battle calc on that territory and see total HP). Hard to say why that is but there is definitely something wrong with that save game as units shouldn't be able to have 0 HP like that.

@justbleh For AI feedback/thoughts, please post in the forum and the primarily AI thread is here: https://forums.triplea-game.org/topic/105/ai-development-discussion-and-feedback. I'm the main AI developer but I haven't really been updating it much recently as I'm focused on adding a few other features to the engine.

**Comment by ssoloff** [comment]
> The Troll in Weathertop actually has 0 HP (you can check by opening battle calc on that territory and see total HP).

Good catch.  Didn't even think to use the Battle Calc that way.  That explains why it takes two rounds to get that particular unit back to 2 HP.

> there is definitely something wrong with that save game

I went all the way back to the beginning of round 1, and I was surprised to see that the units in question (dragon in East Mountains of Angmar, two trolls in Trollfells, and troll in Ettenmoors) all start with 1 HP.  At first, I thought maybe that was SOP for this map, but I started a new game and confirmed the same units have 3, 2, and 2 HP, respectively.

@justbleh I'm not sure what else to try here.  As @ron-murhammer pointed out, there seems to be something wrong with this save game.  Is there any other information you can provide that might be useful?  For example, did you start this game using a different version of TripleA before continuing play in 1.9.0.0.7509?  Or, at any time, did you use the "Save game at this point (BETA)" feature?

**Comment by FrostionAAA** [comment]
Sorry. I don't think that I can contribute to any issue solving here. Its above my skill level. But I can say that the Dragon War map and xml also have self healing multi HP units, in the form of dragons. I haven't seen any problems with them not healing as they should; 1 HP every round and more than 1 HP every round when stationed in cities. But in Dragon War there are never more than 1 unit of each dragon type, so in that way the xml is a bit different.

**Comment by panguitch** [comment]
Same problem in Greyhawk Wars. It doesn't always happen, and I haven't been able to pin down a reason or pattern, but sometimes the map opens with all two-hit units already wounded, and sometimes with the game underway they will all suddenly become wounded.

Look at the attached save game. [15112017.zip](https://github.com/triplea-game/triplea/files/1476540/15112017.zip) Almost every two-hit unit on the map is wounded. Including those that have never fought, like those in the Horned Society, or neutral units, like those in Vesve Forest. Oddly, going back in the game history, it now appears as if they started that way. But they didn't. I was testing to reproduce this error, so I am sure they did not start wounded and only became so later. 

Having a devil of a time trying to recreate with any consistency.

(In case it matters: Greyhawk two-hit units do not heal, unless they are allied with Furyondy and in the same territory as Furyondy's leader.)

**Comment by justbleh** [comment]
I did start the game in version 1.9.0.0.7378 then when I had an issue with it crashing I updated to the pre-release 1.9.0.0.7509.  I did not save in beta feature. Should I (wince) delete all of my saved games uninstall and re-install?

**Comment by ssoloff** [comment]
@justbleh This is sounding more like a bug in the engine rather than a compatibility issue between different versions, so I'm less inclined to think it had anything to do with you starting the game in one version and continuing it in another (especially given @panguitch's comments).

@panguitch Thanks for those observations.  Any additional information you can provide on the repro steps would be greatly appreciated.  In the meantime, I'll try to add some debug code to detect when two-hit units have their HP reduced to see if I can figure out what's triggering the unexpected damage.

**Comment by ssoloff** [comment]
I dumped all the `Change` objects from the Angmar save game history (see [here](https://github.com/triplea-game/triplea/files/1479974/all-history.zip)).  During the repair phase at the end of the first player's (Dwarves) first turn (isolated Dwarves' first round history [here](https://github.com/triplea-game/triplea/files/1479930/dwarves-round-1-history.txt)), nine units are repaired:

```
Unit hits change, hits:IntegerMap:
fortress owned by Dwarves -> 0
wall owned by Dwarves -> 0
wall owned by Dwarves -> 0
wall owned by Dwarves -> 0
wall owned by Dwarves -> 0
wall owned by Dwarves -> 0
wall owned by Dwarves -> 1
wall owned by Dwarves -> 0
wall owned by Dwarves -> 0
, undoHits:IntegerMap:
fortress owned by Dwarves -> 1
wall owned by Dwarves -> 1
wall owned by Dwarves -> 1
wall owned by Dwarves -> 1
wall owned by Dwarves -> 1
wall owned by Dwarves -> 1
wall owned by Dwarves -> 2
wall owned by Dwarves -> 1
wall owned by Dwarves -> 1
```

If you review the earlier changes in the Dwarves' first round history, you'll see that none of the repaired units were involved in combat.  This seems to confirm that those particular units were initialized at the start of the game with non-zero hits.  I found it interesting that not all of these units were initialized to the same non-zero value.  For example, why does only one wall start with 2 hits, while the rest start with 1?

If you look through the full history (search for "Unit hits change"), you'll see similar repairs for other players in the first round, including the non-uniform starting hits.

I've still been unable to reproduce this, but I'll continue to try.

**Comment by ssoloff** [comment]
A weak theory, but I'll just throw it out there...  In #2540 (first included in 1.9.0.0.7395), the Battle Calculator implementations were changed so that they no longer clone the current `GameData` before running the simulations.  Since

* both save games exhibiting this problem use the Hard AI,
* the Hard AI uses the Battle Calculator to generate a ranking for attacking multiple territories, and
* the unexpected damage to the multi-hit units doesn't happen during a Human turn,

I'm wondering if damage that occurs during a simulation somehow makes its way back to the original `GameData` since it is no longer being cloned?  If that's the case, I don't understand why this is so hard to reproduce.  It seems like it should happen often, especially because all multi-hit units seem to be affected when the problem does occur.

@ron-murhammer Can you think of any thing that would invalidate this theory?  I don't want to keep going down this path if it's not viable.

**Comment by prastle** [comment]
https://forums.triplea-game.org/topic/387/world-war-ii-v5-1942-second-edition-battleship-two-hits-issue

Maybe this will help just throwing it in there.

**Comment by ssoloff** [comment]
Thanks, @prastle.  I'm not sure it's related, but I'll keep it in mind.

Also, going back through the history of this issue, I see that @justbleh stated he started his game with 1.9.0.0.7378.  Since the history of that save game suggests the unexpected damage happens on the very first player's turn (i.e. due to the fact that a change was recorded to repair the Dwarves' multi-hit units), it wouldn't be possible for #2540 to have been the root cause at that time.  Oh well, back to the drawing board...

**Comment by panguitch** [comment]
Here's another save game. [17112017.zip](https://github.com/triplea-game/triplea/files/1484128/17112017.zip)  I ran game after game until it happened. Between Great Kingdom's and Nyrond's turns (at some point after Great Kingdom's noncombat movement and Nyrond's political actions) the two-hit units suddenly flipped to wounded.

Happened right before my eyes and I still don't know what triggered it. Grrr.

Note that it's not every two-hit unit on the map. Those that are more remote seem less likely to get wounded. Perhaps it does have something to do with the battle calculator as @ssoloff suggests, since the calculator is less likely to be touching distant/inaccessible units.

**Comment by panguitch** [comment]
And while I'm not certain it can't happen under other AI's, I'm pretty sure I've only seen it under Hard AI.

**Comment by ssoloff** [comment]
Yay!  Using @panguitch's latest scenario (Nyrond: Human, all others: Hard AI), I've been able to reliably reproduce the issue three times.  During each of those repros, I observed the Battle Calculator running and applying hits to multi-hit units.  I still don't know if #2540 is the root cause, but I'm once again highly suspect of it based on the information in the following paragraph.

After a repro, in order for me to reproduce the issue again, I have to exit the game and restart.  I cannot simply leave, return to the main menu, and start a new game.  For some reason, in the latter scenario, the Hard AI is no longer aggressive and doesn't attack.  I don't even observe the Battle Calculator running and applying hits to multi-hit units, as I did above.  When there are no attacks, and the Battle Calculator doesn't run, I don't observe the unexpected hits by the time it's Nyrond's turn.  Not sure if there's some state being left in memory that causes the Hard AI to change its behavior when not fully exiting the application.

Regardless, now that I can repro, I will revert #2540 locally to quickly see if that's the root cause, and then proceed from there.

**Comment by ssoloff** [comment]
I'm now pretty sure that #2540 is the root cause.  I couldn't easily revert it due to the structural changes in that PR, so instead I tested 1.9.0.0.7391 (release prior to #2540) and 1.9.0.0.7395 (first release with #2540).  I confirmed that the unexpected hits do not occur in 7391 but do occur in 7395.

I also confirmed that the passive behavior of the Hard AI after leaving and re-starting a game is not present in 7391, so that's probably due to #2540, as well.

I'll see if I can just fix up #2540 under the assumption not cloning the `GameData` is the problem.  I'd prefer not to revert the entire PR because it contains several good changes

**Comment by ssoloff** [comment]
A naïve cloning of `GameData` per Battle Calculator run won't work because it creates too much pressure on the heap.  The Hard AI ends up performing almost 1,000 runs for the first player in the Greyhawk map.  CPU usage quickly spikes while the GC tries to keep up with all the `GameData`-related objects being created and destroyed.

The original implementation overcame this by using a single `GameData` instance per calculator instance and simply reverting all changes at the end of each run.  This resulted in only about 10 `GameData` instances being created in the same scenario as above.  Restoring this capability does not appear to be trivial.

There are a few other issues I found while playing around with the Battle Calculator code:

1. `ProAI` uses a single (`static`) instance of `ConcurrentOddsCalculator`.  If the `cancel` flag for that instance is set, the calculator will never run another simulation.  I believe this is why I have to exit the application to start another game, as described above, in order prevent the Hard AI from becoming passive.  Even though the user or AI do not explicitly ever cancel the Battle Calculator, there's some code that seems to automatically mark it as canceled at the end of an operation.
1. `ProAI#purchase()` makes a copy of the `GameData` and manipulates it.  It expects the `ConcurrentOddsCalculator` to use this modified copy, but that doesn't appear to happen.
1. For `ConcurrentOddsCalculator#cancel()` to work as intended, it really needs to cancel the individual tasks submitted to the executor.  Setting the `cancelled` flag simply stops new tasks from being submitted; it will not interrupt tasks that have already been submitted even if they are still queued for execution.
1. There's a race condition in `ConcurrentOddsCalculator#calculate()`.  The `OddsCalculatorParameters` run count is set to 1 before creating each task and then reset back to the original count after all tasks have been submitted.  However, since each task keeps a reference to the original `OddsCalculatorParameters` instance, they may see the original run count instead of 1, as intended.  I'm not sure this matters because it doesn't appear that `OddsCalculator#doSimulation()` ever uses the run count, so it may be sufficient to simply remove the `setRunCount()` calls.
1. There are two calls to `doSimulation()` for each run in `OddsCalculator#calculate()`.  The results of the first call are ignored.  I'm not sure if this was intended as some kind of "warm-up" run.  If not, it's simply doubling the number of runs.

So, with regret because I really like the direction #2540 was going, I think we just need to revert it for now.  Fixing the correctness and performance issues might take several days.  Given that this is a live bug affecting all Hard AI games with multi-hit units (and possibly other unexpected changes we have yet to discover), I think a quick revert and new release is warranted.

**Comment by ssoloff** [comment]
@justbleh @panguitch This should be fixed in 1.9.0.0.7559.  Please re-open this issue if you still observe the problem.
